pkg/controllers/hub/trafficmanagerprofile/controller.go

```diff
@@ -0,0 +1,271 @@
+	if getErr != nil {
+		if !azureerrors.IsNotFound(getErr) {
+			klog.ErrorS(getErr, "Failed to get the profile", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return ctrl.Result{}, getErr
+		}
+		klog.V(2).InfoS("Azure Traffic Manager profile does not exist", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+	} else {
+		existingSpec := convertToTrafficManagerProfileSpec(&getRes.Profile)
+		if equality.Semantic.DeepEqual(existingSpec, profile.Spec) {
+			// skip creating or updating the profile
+			klog.V(2).InfoS("No profile update needed", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return r.updateProfileStatus(ctx, profile, getRes.Profile, nil)
+		}
+	}
+
+	res, updateErr := r.ProfilesClient.CreateOrUpdate(ctx, r.ResourceGroupName, azureProfileName, generateAzureTrafficManagerProfile(profile), nil)
+	if updateErr != nil {
+		if !errors.As(updateErr, &responseError) {
+			klog.ErrorS(updateErr, "Failed to send the createOrUpdate request", "trafficManagerProfile", profileKObj, "azureProfileName", azureProfileName)
+			return ctrl.Result{}, updateErr
+		}
+		klog.ErrorS(updateErr, "Failed to create or update a profile", "trafficManagerProfile", profileKObj,
+			"azureProfileName", azureProfileName,
+			"errorCode", responseError.ErrorCode, "statusCode", responseError.StatusCode)
+	}
+	return r.updateProfileStatus(ctx, profile, res.Profile, updateErr)
+}
+
+func convertToTrafficManagerProfileSpec(profile *armtrafficmanager.Profile) fleetnetv1alpha1.TrafficManagerProfileSpec {
+	if profile.Properties != nil && profile.Properties.MonitorConfig != nil {
+		var protocol fleetnetv1alpha1.TrafficManagerMonitorProtocol
+		if profile.Properties.MonitorConfig.Protocol != nil {
+			protocol = fleetnetv1alpha1.TrafficManagerMonitorProtocol(*profile.Properties.MonitorConfig.Protocol)
+		}
+		return fleetnetv1alpha1.TrafficManagerProfileSpec{
+			MonitorConfig: &fleetnetv1alpha1.MonitorConfig{
+				IntervalInSeconds:         profile.Properties.MonitorConfig.IntervalInSeconds,
+				Path:                      profile.Properties.MonitorConfig.Path,
+				Port:                      profile.Properties.MonitorConfig.Port,
+				Protocol:                  &protocol,
+				TimeoutInSeconds:          profile.Properties.MonitorConfig.TimeoutInSeconds,
+				ToleratedNumberOfFailures: profile.Properties.MonitorConfig.ToleratedNumberOfFailures,
+			},
+		}
+	}
+	return fleetnetv1alpha1.TrafficManagerProfileSpec{}
+}
+
+func (r *Reconciler) updateProfileStatus(ctx context.Context, profile *fleetnetv1alpha1.TrafficManagerProfile, azureProfile armtrafficmanager.Profile, updateErr error) (ctrl.Result, error) {
+	profileKObj := klog.KObj(profile)
+	if updateErr == nil {
+		// azureProfile.Properties.DNSConfig.Fqdn should not be nil
+		if azureProfile.Properties != nil && azureProfile.Properties.DNSConfig != nil {
+			profile.Status.DNSName = azureProfile.Properties.DNSConfig.Fqdn
```

Comment:
Is the Fqdn field filled by Azure resource provider? Will be done automatically in `CreateOrUpdate` function?

Reply:
Yes, it's returned by the Azure resource Provider as the output in the CreateOrUpdate response.

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/trafficmanager/armtrafficmanager#DNSConfig

```
type DNSConfig struct {
	// The relative DNS name provided by this Traffic Manager profile. This value is combined with the DNS domain name used by
	// Azure Traffic Manager to form the fully-qualified domain name (FQDN) of the
	// profile.
	RelativeName *[string](https://pkg.go.dev/builtin#string)

	// The DNS Time-To-Live (TTL), in seconds. This informs the local DNS resolvers and DNS clients how long to cache DNS responses
	// provided by this Traffic Manager profile.
	TTL *[int64](https://pkg.go.dev/builtin#int64)

	// READ-ONLY; The fully-qualified domain name (FQDN) of the Traffic Manager profile. This is formed from the concatenation
	// of the RelativeName with the DNS domain used by Azure Traffic Manager.
	Fqdn *[string](https://pkg.go.dev/builtin#string)
}
```